### ISSUE TITLE
try borg with pypy (and cpyext) - does it work?

### ISSUE THREAD

**Issue by ThomasWaldmann** [borgbackup/borg#1755]

**Comment by ThomasWaldmann** [comment]
Tried borg with current pypy3.5 v5.9:
- borg seems to work (tried some basic commands shortly)
- tests have quite some issues: integrity errors, crashing test workers with pytest-xdist, slow msgpack, ...

**Comment by enkore** [comment]
5.8 didn't work due to some missing APIs (PyMemoryView_FromMemory).

**Comment by ThomasWaldmann** [comment]
did a quick test with pypy3.5-6.0.0 (installed on stretch64 vagrant VM via pyenv):
- llfuse does not build
- without fuse, borg builds and tests run, but quite some test failures

**Comment by ThomasWaldmann** [comment]
one could do a new test using pyfuse3 after #4122 is closed.

**Comment by zougloub** [comment]
Briefly tested:
```sh
pypy3 -m pip install --user git+https://github.com/borgbackup/borg.git
```

Findings:
- Seemed to work, I tested rcreate, create, rlist, list, mount
- Message `Using a pure-python msgpack! This will result in lower performance.`

**Comment by ThomasWaldmann** [comment]
@zougloub Ah, interesting, thanks for testing!

The pure-python msgpack is somewhat expected and maybe not even that slow as it will also be accelerated by pypy.

Wondering whether msgpack could be made working with the faster C code, considering that it also seemed to cope somehow with borg and its C/Cython code.

**Comment by ThomasWaldmann** [comment]
Retested with **pypy3.11 v7.3.23** (homebrew, macOS/arm64) against borg2 master (2.0.0b23.dev):

- borg **builds fine** via cpyext (`pip install -e .`) — all dependencies have pypy wheels or build from source (borghash, backports-zstd, argon2-cffi; msgpack falls back to pure-python as before).
- After some small borg-side workarounds (branch `try-pypy`, PR upcoming), the **full test suite passes** on pypy: 2056 passed, 1 xfailed — and the complete workflow (repo-create, create, list, extract, check, delete) works correctly.

Found four pypy bugs/problems (macOS) that needed workarounds in borg:

1. pypy's `fcntl` module does not expose `F_FULLFSYNC` → every command that writes crashed in `platform.darwin.fdatasync`. Workaround: fall back to the numeric constant (51).
2. pypy's `hmac.digest` / `_hashlib` only accepts `bytes` (neither `memoryview` nor `bytearray`) → `borg create` crashed in `hmac_sha256` chunk id hashing. Workaround: coerce to bytes on pypy.
3. pypy's `os.link(..., follow_symlinks=False)` raises `EINVAL`, although `os.supports_follow_symlinks` claims support → all hardlink extraction failed. Workaround: emulate the secure follow_symlinks=False behaviour (plain link for regular targets, an equal symlink when the hardlink target is a symlink).
4. **Severe memory leak at scale**: the Cython chunkers return `memoryview(bytes_chunk)`; under pypy, a memoryview over a C-API-created bytes object pins the buffer via cpyext, and if the consumer drops it without calling `.release()`, the memory is *never* reclaimed — not even by `gc.collect()`. `borg create` of a 20 GB dataset grew to a **43 GB** footprint and got killed by the OS. Fix: explicitly `.release()` the memoryview in all consumers of chunker output; memory then stays flat (~160 MB for the same create). This was invisible on small datasets.

One behavioural difference: setting undeclared attributes on Cython cdef class instances (e.g. `Item`) is not blocked under pypy, so typos are silently ignored instead of raising AttributeError (test xfailed).

**Performance** (same machine, CPython 3.11 vs pypy3.11):

- small dataset (500 MB / 5500 files): pypy ~3-4x slower across create/extract/check — per-process JIT warmup dominates.
- big dataset (20 GB / 5000 files, chacha20-poly1305 + lz4): pypy ~2x slower on data-heavy paths (create cold: 151 s vs 73 s ≈ 132 vs 274 MB/s; extract: 74 s vs 33 s; check: 22 s vs 10 s), ~3x on the metadata-heavy files-cache-hit create (106 s vs 34 s — pure-python msgpack territory).

The hot paths are Cython/C extensions running through pypy's cpyext emulation, so pypy support is a compatibility story, not a speed win.

Conclusion: it works. Closing — the remaining issues are pypy bugs, and the expected slowness makes pypy a non-goal beyond basic compatibility.

**Comment by ThomasWaldmann** [comment]
#9976 after some issues were fixed in pypy, this is the PR that makes borg work with pypy.

Also, it is quite slow. borg does a lot in C (cython) code and whenever pypy interfaces to C code, it gets slow, very slow. PyPy also makes borg's Python code much faster, but that has a much smaller impact, because all the heavy lifting is done in C code / C libraries.
